src/Language/Haskell/Liquid/Constraint/Types.hs

```diff
@@ -233,30 +233,55 @@ elemHEnv x (HEnv s) = x `S.member` s
 -- | Helper Types: Invariants --------------------------------------------------
 --------------------------------------------------------------------------------
 
-type RTyConInv = M.HashMap RTyCon [SpecType]
-type RTyConIAl = M.HashMap RTyCon [SpecType]
+data RInv = RInv { _rinv_args :: [RSort]
+                 , _rinv_type :: SpecType
+                 }
+
+type RTyConInv = M.HashMap RTyCon [RInv]
+type RTyConIAl = M.HashMap RTyCon [RInv]
 
 --------------------------------------------------------------------------------
 mkRTyConInv    :: [F.Located SpecType] -> RTyConInv
 --------------------------------------------------------------------------------
-mkRTyConInv ts = group [ (c, t) | t@(RApp c _ _ _) <- strip <$> ts]
+mkRTyConInv ts = group [ (c, RInv (go ts) t) | t@(RApp c ts _ _) <- strip <$> ts]
   where
-    strip      = fourth4 . bkUniv . val
+    strip = fourth4 . bkUniv . val
+    go ts | generic (toRSort <$> ts) = []
+          | otherwise                = toRSort <$> ts 
+
+    generic ts = let ts' = L.nub ts in
+                 all isRVar ts' && length ts' == length ts 
 
 mkRTyConIAl :: [(a, F.Located SpecType)] -> RTyConInv
 mkRTyConIAl    = mkRTyConInv . fmap snd
 
 addRTyConInv :: RTyConInv -> SpecType -> SpecType
-addRTyConInv m t@(RApp c _ _ _)
-  = case M.lookup c m of
+addRTyConInv m t
+  = case lookupRInv t m of
       Nothing -> t
-      Just ts -> L.foldl' conjoinInvariantShift  t ts
-addRTyConInv _ t
-  = t
+      Just ts -> L.foldl' conjoinInvariantShift t ts
+
+lookupRInv :: SpecType -> RTyConInv -> Maybe [SpecType]
+lookupRInv (RApp c ts _ _) m 
+  = case M.lookup c m of
+      Nothing   -> Nothing
+      Just invs -> Just (catMaybes $ goodInvs ts <$> invs)
+lookupRInv _ _
+  = Nothing 
+
+goodInvs :: [SpecType] -> RInv -> Maybe SpecType
+goodInvs _ (RInv []  t) 
+  = Just t 
+goodInvs ts (RInv ts' t)
+  | ts' == (toRSort <$> ts)
```

Comment:
This will only work if the `Type`s match exactly. What if I have

```
invariant {v: Ptr a | ptrSize v >= 0}
invariant {v: Ptr Word16 | ptrSize v = 2}
```

A `Ptr Word16` would then only get the 2nd invariant, but it should get both. Furthermore, a `Ptr b` would probably get neither right?